net/sched/sch_gkprio.c

```diff
@@ -0,0 +1,288 @@
+			   struct sk_buff **to_free)
+{
+	struct qdisc_skb_head *your_queue;
+	struct sk_buff *pkt;
+	int wlen;
+	u16 your_prio, lp;
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+
+	wlen = skb_network_offset(skb);
+
+	switch (tc_skb_protocol(skb)) {
+	case htons(ETH_P_IP):
+		wlen += sizeof(struct iphdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+
+		your_prio = ipv4_get_dsfield(ip_hdr(skb)) >> 2;
+		break;
+
+	case htons(ETH_P_IPV6):
+		wlen += sizeof(struct ipv6hdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+
+		your_prio = ipv6_get_dsfield(ipv6_hdr(skb)) >> 2;
+		break;
+
+	default:
+		your_prio = 0;
+		break;
+	}
+
+	/* Check to update highest and lowest priorities. */
+
+	if (your_prio > q->highest_prio)
+		q->highest_prio = your_prio;
+
+	if (your_prio < q->lowest_prio)
+		q->lowest_prio = your_prio;
+
+	your_queue = &q->prio_array[your_prio];
+
+	if (sch->q.qlen < q->max_limit) {
+
+		sch->q.qlen++;
+		return __qdisc_enqueue_tail(skb, sch, your_queue);
+	}
+
+	/* Look to lowest priority queue to check if ingress packet can be
+	 * accomodated. Otherwise no choice but to drop ingress packet.
+	 */
+
+	lp = q->lowest_prio;
+	if (lp != your_prio) {
+		struct qdisc_skb_head *lpq = &q->prio_array[lp];
+
+		/* Drop the packet at the tail of lowest_priority queue. */
+		pkt = lpq->tail;
+		if (likely(pkt)) {
+			lpq->tail = pkt->prev;
+			lpq->qlen--;
+			if (!lpq->tail)
+				lpq->head = NULL;
+			pkt->prev = NULL;
+		}
+
+		__qdisc_drop(pkt, to_free);
+		qdisc_qstats_drop(sch);
+
+		/* Update lowest priority field. */
+		if (lpq->qlen == 0)
+			q->lowest_prio = calc_new_low_prio(q);
+
+		return __qdisc_enqueue_tail(skb, sch, your_queue);
+	}
+
+	return qdisc_drop(skb, sch, to_free);
+drop:
+	qdisc_drop(skb, sch, to_free);
+	return NET_XMIT_SUCCESS | __NET_XMIT_BYPASS;
+}
+
+static struct sk_buff *gkprio_dequeue(struct Qdisc *sch)
+{
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+	struct sk_buff *skb;
+	u16 hp = q->highest_prio;
+
+	struct qdisc_skb_head *hpq = &q->prio_array[hp];
+
+	skb = __qdisc_dequeue_head(hpq);
+	if (unlikely(!skb))
+		return NULL;
+
+	sch->q.qlen--;
+	qdisc_qstats_backlog_dec(sch, skb);
+	qdisc_bstats_update(sch, skb);
+
+	/* Update highest priority field. */
+	if (hpq->qlen == 0)
+		q->highest_prio = calc_new_high_prio(q);
+
+	return skb;
+}
+
+static int gkprio_init(struct Qdisc *sch, struct nlattr *opt)
+{
+	int prio;
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+
+	if (!opt) {
+		q->max_limit = qdisc_dev(sch)->tx_queue_len;
+	} else {
+		struct tc_gkprio_qopt *ctl = nla_data(opt);
+
+		if (nla_len(opt) < sizeof(*ctl))
+			return -EINVAL;
+
+		q->max_limit = ctl->limit;
+	}
+
+	/* Initialise 64 queues, one for each possible priority. */
+	for (prio = 0; prio < MAX_PRIORITY; prio++)
+		qdisc_skb_head_init(&q->prio_array[prio]);
+
+	q->highest_prio = 0;
+	q->lowest_prio = MAX_PRIORITY - 1;
+	return 0;
+}
+
+static int gkprio_dump(struct Qdisc *sch, struct sk_buff *skb)
```

Comment:
Shouldn't we be providing more information about the state of the GK priority queue in this function, not just its size limit?